### PR TITLE
Refs #373 -- Added Model._is_pk_set().

### PR DESCRIPTION
**Pull request by csirmazbendeguz**
**[GSoC-2024]**

# Trac ticket number
ticket-373

# Branch description
Added the `Model._is_pk_set()` function as suggested by @charettes and @LilyFoote (https://github.com/django/django/pull/18056#discussion_r1604282430).

It's a simple utility function for checking if the primary key has been set on an object.

It's a pre-requisite for composite primary keys (https://github.com/django/django/pull/18056).

This function will get modified by the [composite primary keys patch](https://github.com/django/django/pull/18056/files#diff-1c8b882c73bfda668d5451d4578c97191b0ebc0f088d0c0ba2296ab89b428c44R672).

# Checklist
- [X] This PR targets the `main` branch. 
- [X] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [X] I have checked the "Has patch" ticket flag in the Trac system.
- [X] I have added or updated relevant tests.
- [X] I have added or updated relevant docs, including release notes if applicable.
- [X] I have attached screenshots in both light and dark modes for any UI changes.

### CODE REVIEW

**Comment by nessita** [comment]
Thank you @csirmazbendeguz for this PR. I have read the referenced comment and while I understand the rationale, I have some concerns about the chosen name (though I acknowledge that this PR follows the proposed name).

When I read this, I thought that `is_set` was referring to "isinstance(..., set)", i.e. that it answers "is this a set" (instead of the current semantic which is "is this set"). Risking some bikeshedding, I was thinking that perhaps a better name would be:
* is_defined
* value_set
* is_value_set
* has_value_set

cc/ @LilyFoote @charettes

**Comment by csirmazbendeguz** [comment]
Thanks @nessita! It's not bikeshedding at all, I submitted this PR separately so it's easier to review and so we can have a discussion about it without worrying about all the other details of composite primary keys.

I like `is_value_set`. 👍

**Comment by shangxiao** [comment]
> When I read this, I thought that `is_set` was referring to "isinstance(..., set)", i.e. that it answers "is this a set" (instead of the current semantic which is "is this set"). Risking some bikeshedding, I was thinking that perhaps a better name would be:

I haven't been part of the composite thread but when I saw this my first thought was similar: "has the value on this field been set?" … though when I glanced at the `is_value_set()` I still thought it had the same semantic just named differently 😅 

There are 2 hard problems in computer science:

1. Naming
2. Caching things
3. Off-by-1 errors

**Comment by csirmazbendeguz** [comment]
Yes it feels a little awkward.
This function is supposed to be an abstraction for checking if the pk has been set on an object.
So maybe:
- `has_null`?
- `is_null`?

File: django/db/models/base.py
Comment (sarahboyce) [inline django/db/models/base.py:1098]:
```suggestion
        if not meta.pk.is_value_set(pk_val):
```
(maybe?)

**Comment by csirmazbendeguz** [comment]
Thanks @sarahboyce ! I adjusted it.

If we can't come up with a good name then maybe this is not the best abstraction, in which case I would be open to alternatives.

1. Since composite primary keys are tuples, checking if `pk is None` is not enough. We need to check if the tuple contains null values instead.
2. Since there are multiple instances of pk null checks, it makes sense to put this in a function. The question is, where do we place this function and what do we call it?

**Review by LilyFoote — APPROVED** [review]
`is_null` sounds good to me (though I don't have a problem with the other names either).

**Comment by nessita** [comment]
Thank you everyone for engaging in the naming conversation. I agree that `is_null` is a good option! +1 to that choice.

**Comment by csirmazbendeguz** [comment]
I do wonder if this could be mistaken as the getter for `Field.null`.
Since this function is for checking the pk value, we could also include "pk" in the name.
- `is_none`
- `is_pk_none`
- `is_null_value`
- `is_pk_null_value`
- `is_nullval`
- `is_pk_nullval`
- `contains_null`
- `pk_contains_null`
- `contains_none`
- `pk_contains_none`
- `has_none`
- `pk_has_none`

**Review by sarahboyce — APPROVED** [review]
Agreed, given the context and what it is used for, also like `is_pk_value_set`

**Comment by nessita** [comment]
After a more throrough review I have posted this in PR #18056:

> I'm honestly confused about why we would put the "responsibility" of knowing about a PK value in a `Field` and not in the `Model` instance. More so, the new method as proposed does not use anything from the `Field` class (and I think it will not either in the `CompositePrimaryKey` override).
> Could we do something like this for `ModelBase` (hand wavy untested code)?

* for `main`:
```python
    def is_pk_set(self):
        return self.pk is not None
```
* this branch:
```python
    def is_pk_set(self):
        if isinstance(self.pk, CompositePrimaryKey):
            return all(value is not None for value in self.pk)
        return self.pk is not None
```

**Comment by csirmazbendeguz** [comment]
Yes @nessita , that could work. I agree it looks more readable as a `Model` function.

**Comment by LilyFoote** [comment]
One theoretical downside of the `Model` function I just thought of is that third-party fields can't opt-in to custom handling here. That said, I can't think of a third-party field that would want to do that.

**Comment by csirmazbendeguz** [comment]
> One theoretical downside of the `Model` function I just thought of is that third-party fields can't opt-in to custom handling here. That said, I can't think of a third-party field that would want to do that.

Yes, I also brought this up on Discord, this is called the _open-closed principle_. I agree with @nessita though, in this case the `Model` function is more intuitive, which is more important than extensibility.

**Review by nessita — COMMENTED** [review]
Thank you @csirmazbendeguz! I have thoroughly reviewed this proposal. I think it looks good but I do have a few questions:
1. Shall the `if self.pk is None:` in `__hash__` (db/models/base.py:604) be replaced as well?
2. Same question for these in the same file:
  * `if obj.pk is None:` in `_prepare_related_fields_for_save`
  * `if obj and obj.pk is None:` in the same method `_prepare_related_fields_for_save`
  * `if self.pk is None:` in `delete`
  * `if self.pk is None` in `prepare_database_save`
  * `model_class_pk is not None` which comes from `model_class_pk = self._get_pk_val(model_class._meta)` in `_perform_unique_checks`
  * `self.pk is not None` in `_perform_date_checks`

I see more matches in other files (when running `grep -nR "pk.*None"`), I know not all of them are suitable to use the helper but could you please confirm if you have reviewed each of these matches to evaluate replacement on a case by case basis?

Thank you!

**Comment by csirmazbendeguz** [comment]
Thanks @nessita ⭐, you're completely right. My bad.
I reviewed all cases and added the function where it makes sense.

**Comment by nessita** [comment]
> Thanks @nessita ⭐, you're completely right. All of these should be replaced. My bad. I reviewed all cases and added the function where it makes sense.

Thank you so much for the updates. I have grepped again after fetching the latest changes and I do have doubts about these:

```
db/models/fields/related_descriptors.py:515:            if related_pk is None:
db/models/fields/related_descriptors.py:1084:            if instance.pk is None:
db/models/fields/related.py:1970:        return [] if obj.pk is None else list(getattr(obj, self.attname).all())
db/models/query_utils.py:215:                if instance.pk is None and self.field.generated:
db/models/options.py:302:        if self.pk is None:
forms/models.py:938:            if obj.pk is None:
forms/models.py:1106:        if self.instance.pk is not None:
contrib/contenttypes/forms.py:34:        if self.instance is None or self.instance.pk is None:
```

(When thinking about composite PKs I think these may also need to "be able" to check whether a single or composite PK is None?)

Could you please confirmed if you have reviewed those and if there are reasons for not to replace them? Thanks again!

**Comment by csirmazbendeguz** [comment]
Yes, I reviewed them, let me list the reasons why I didn't replace:

`
db/models/fields/related_descriptors.py:515:     composite OneToOneField is not supported yet
db/models/fields/related_descriptors.py:1084:   composite ManyToManyField is not supported yet
db/models/fields/related.py:1970:     composite ManyToManyField is not supported yet
db/models/query_utils.py:215:     composite generated fields are not possible
db/models/options.py:302:      this is correct, it refers to the Options.pk, not the Model.pk
forms/models.py:938:        as per the comment above, pk is None if invalid POST data or if the object was already deleted, so I think this is not 100% related?
forms/models.py:1106    I wasn't not sure about this, looking at it closely it makes sense to replace it here
contrib/contenttypes/forms.py:34   I wasn't sure about this, I think it makes sense here too
`

**Comment by collinanderson** [comment]
Does this or another patch need a backward compatible change documentation? Now that people can no longer reliably check if `obj.pk is None` and instead need to call `obj._is_pk_set()`?

Or is it possible to preserve the `obj.pk is None` behavior for composite fields so django apps don't need to make code changes to support composite fields?

**Comment by csirmazbendeguz** [comment]
@collinanderson
It's not possible to get `obj.pk is None` working with composite primary keys.
That said it's not a backwards incompatible change, and `obj.pk is None` is still reliable.
Since composite primary keys will be a new feature, `obj._is_pk_set()` will only be necessary if you use composite primary keys.

**Comment by collinanderson** [comment]
> obj._is_pk_set() will only be necessary if you use composite primary keys.

Right, but there may be generic 3rd party apps that work on any model (like admin extensions), and they'll need to switch their code in order to support these new types of models.

**Comment by csirmazbendeguz** [comment]
Yes @collinanderson, that's correct. We could add this to the release notes maybe? @nessita wdyt?

**Comment by csirmazbendeguz** [comment]
> It's not possible to get obj.pk is None working with composite primary keys.

Let me elaborate on this. It's not that it's impossible, it's that it would lead to weird behaviour.

```
>>> obj.pk = (1, None)
>>> obj.pk
None
```

`pk` would need to return `None` if any of the fields is `None`.

**Comment by collinanderson** [comment]
Right, yeah I'm not super familiar with how composite fields would work, but it could also be surprising to people that:

```
>>> obj.pk = (1, None)
>>> obj._is_pk_set()
False
```

It's weird/awkward in either case, right?

**Comment by csirmazbendeguz** [comment]
I suppose the difference is `obj.pk` is a public API, but `obj._is_pk_set()` is not.

**Comment by csirmazbendeguz** [comment]
@collinanderson, it's a good idea but we can't prevent setting `pk` to `None`, since Django itself sets it to `None` on deletion (see [/django/db/models/deletion.py:517](https://github.com/django/django/blob/43cdfa8b20e567a801b7d0a09ec67ddd062d5ea4/django/db/models/deletion.py#L517)).

`obj.pk = (1, None)` is just one example, it's also possible to instantiate a model without any value at all in which case `obj.pk` will be `(None, None)`.

```
>>> obj = Foo()
>>> obj.pk
(None, None)
```

Or instantiate it with only one `pk` field.

```
>>> obj = Foo(foo=1)
>>> obj.pk
(1, None)
```

It's the same pattern as regular primary keys.

**Comment by nessita** [comment]
> Yes @collinanderson, that's correct. We could add this to the release notes maybe? @nessita wdyt?

I'm a little torn by this, let me explain where my head is at:
* This PR started as a refactor to isolate the logic for checking if a Model's PK is set or not
* My understanding is that this helper would be used internally only
* Later on, in the Composite PK branch, this helper would be extended to "learn" how to handle composite PKs

The point raised by @collinanderson makes sense in that I can see how third party apps would benefit from using "a blessed method to check if a PK, compound or not, is set". With this information, I see two options:

* We leave this work as is, considering `_is_pk_set` an internal, undocumented method, that is used to facilitate the progression of Composite PKs. Then, once composite PK is ready/landed, we revisit this goal of allowing third party apps to easily and robustly check for set PKs. Pro: we would expose a new method we know it works since it has already a IRL proof of usage with Composite PKs. Con is that this could delay adoption of the feature for third party apps.

Or:

* We create a new feature ticket for providing a public and documented method to check if a PK is set and fully flesh the design, API, name, docs, etc in the ticket, for then repurpose this PR to solve that ticket. Pro: API for public consumption is ready faster and earlier. Con: this may be a "slower" path for composite PK feature as a whole.

@LilyFoote @charettes @carltongibson Would you have an opinion on the above?

**Comment by LilyFoote** [comment]
I think we can land this now as an internal API and also open a new ticket about a public API, which can progress independently of the composite PK work.

**Comment by csirmazbendeguz** [comment]
We can rename this `_is_pk_set` -> `is_pk_set`, add some docs and that's it right? I can open a new ticket if it's necessary, but I'm not sure if there's much else to flesh out.

I assume it's going to take a while for maintainers to review the composite primary keys PR, so we might as well make this function a public API.

**Comment by carltongibson** [comment]
I'm struggling to see an easy way around the issue here. (Reminder to self: Issue is that I can't keep using `obj.pk` to test whether the object has been saved.) 

Elsewhere I've used `obj._state.adding` but I don't think documenting that is the way forward. 

So, I'd suggest just adding `is_pk_set()` a public documented property, and telling folks code will need to adopt to that if they want to support models with composite keys. It's `Refs #373 ...` — surely there has to be **some** disruption 😅 

@collinanderson Would that be OK for you do you think? (I don't want to have missed your concern.)

Having read the thread here I'd say, add the new method as public to begin. I've likely missed the reason for the initial private phase there though. (Please continue if that's so.)

**Comment by csirmazbendeguz** [comment]
@carltongibson , I agree, but note that

> Issue is that I can't keep using obj.pk to test whether the object has been saved.

This is close to the truth, but it can also used for other purposes, e.g. to check if `default` should be used for the `pk` field.
It's not a matter of `obj._state.adding` vs `is_pk_set()`, the two are sometimes used together.

On `obj.save()`:

| `obj._state.adding` | `obj.is_pk_set()` | result                                             |
| -------------------- | ---------------- | ------------------------------------- |
| `True`                     | `True`                | `obj` is created with provided `pk` |
| `True`                     | `False`               | `obj` is created with default `pk`    |
| `False`                    | `True`                | `obj` is updated                             |
| `False`                    | `False`               | `obj` is cloned                                |

**Comment by csirmazbendeguz** [comment]
Thanks for this @collinanderson !

It's good that you brought this up, because some backends do depend on this. Here's an example from [microsoft/mssql-django](https://github.com/microsoft/mssql-django/blob/cdad07b8a3a06d6127b38bb03010eadf81436518/mssql/functions.py#L310).

There is a special case for `obj.pk = None` [composite.py:30](https://github.com/django/django/pull/18056/files#diff-96b2a1412a76cbb213805a53e97c58c828f046adaa65fd00030646b5c251827dR30), this is to retain compatibility with regular `pk`s and to avoid changes to deletes. I agree with the sentiment that backward compatibility is good.

To set the defaults, there's `get_pk_value_on_save` [composite.py:105](https://github.com/django/django/pull/18056/files#diff-96b2a1412a76cbb213805a53e97c58c828f046adaa65fd00030646b5c251827dR105).

**Comment by csirmazbendeguz** [comment]
> `pk` would need to return `None` if any of the fields is `None`. It's not ideal IMHO.

I was against doing this at first, but now that @collinanderson made me think about it, I can't come up with any downsides other than "it looks like a hack and I don't like it".

- On one hand, `(None, None)` or `(123, None)` is more consistent with the idea of composite primary keys being represented as tuples.
- On the other hand, `None` gives us more compatibility.

So I would like to ask everyone to give me their opinion. What do you think? Which one is better?

**Comment by collinanderson** [comment]
Could you ever have a composite-primary-key where one or both of the columns are nullable? In which case my idea of returning `None` would probably be bad. Too much "magic" can bad.

**Comment by csirmazbendeguz** [comment]
> Could you ever have a composite-primary-key where one or both of the columns are nullable?

A primary key column cannot be nullable (composite or not).

**Comment by LilyFoote** [comment]
I think if we do special-case here we should only do so for the case where every field is `None`, not when some are set.

Beyond that, I can see arguments either way and I'm not sure which is better.

**Comment by csirmazbendeguz** [comment]
Thanks @LilyFoote !

If returned `None` only if `obj.pk == (None, None)` that wouldn't be enough because we would still need to add `Model.is_pk_set()`.

**Comment by carltongibson** [comment]
> So I would like to ask everyone to give me your opinion. What do you think? Which one is better?

I can't say definitively but, I'm leaning to the _less magic, more explicit_ tuple response, even for the `(None, None)` cases, despite that needing the new API (which probably should be used anyway right...? — If I followed that last twist in the conversation right.) 

I can't shake the feeling that if we convert to `None` just in the one case (or at all TBH) we'll be saving up some horrible need for a workaround much later. 

🤷

**Comment by csirmazbendeguz** [comment]
> despite that needing the new API (which probably should be used anyway right...? — If I followed that last twist in the conversation right.

What I meant is if we're going to move this `is_pk_set` logic to the new descriptor (`CompositeAttribute.__get__`) we gotta check if _any_ of the fields is None, not only if all the fields are None.

**Comment by charettes** [comment]
Thanks for the ping and sorry for the late answer. I'll try to summarize all the points brought so far to

1. Managing to preserve `pk is None` behavior would reduce backward compatibility concerns at the cost of possibly unintuitive magic (if any member of the composition are `None`)
2. Should we have a public of private `_?is_pk_set` method (assuming we can agree on a name)

While I think that we should strive to reduce the backward compatibility burden imposed by the introduction of composite primary keys I don't think that focusing on making `pk` return `None` in this case is worthy for two reasons.

The first one being that one way or the other third-party app that ought to support composite primary keys will have to be adapted to expect `pk` to return a `tuple` under some conditions, in this sense `pk is None` is just one pattern that will require adjustments.

The second one is that `pk is None` has always been an _implicit_ way through convention to get an idea of how to treat a `Model` instance. A pattern, just like checking for `_state.adding`, that has been replicated as a symptom of a lack of a blessed way to identify the persistence state of a `Model` instance. In this sense I think that we should see the introduction of composite primary keys as an opportunity to direct the ecosystem towards a blessed way of doing things instead of encouraging this convention further.

I really like the table that @csirmazbendeguz put together in https://github.com/django/django/pull/18450#issuecomment-2293279873 as I believe it is the closest to what I would expect a proper public API to look like if we want to commit to one; something that resolves both the `pk is None` and `_state.adding` ambiguity at the same time instead of focusing solely on the notion of `pk`.

The approach I would favor here is to start with a `_is_pk_set` API merged into `main` to validate the approach and take an approach where we document its existence but mention it is likely to be superseded eventually through a deprecation period. If we're able to come up with a solution that supersedes both `pk is None` and `_state.adding` checks in a nice way before its release in a distinct work stream it won't be an issue otherwise we'll get to learn through before committing to a solution.

File: django/db/models/base.py
Comment (nessita) [inline django/db/models/base.py:1100]: I think this can now be removed from this line and added further down in line 1132 (as of this PR) right before the call `self._do_update`. Specifically I mean:
```diff
diff --git a/django/db/models/base.py b/django/db/models/base.py
index 5187e9e35c..5b819b1406 100644
--- a/django/db/models/base.py
+++ b/django/db/models/base.py
@@ -1097,7 +1097,6 @@ class Model(AltersData, metaclass=ModelBase):
                 if f.name in update_fields or f.attname in update_fields
             ]
 
-        pk_val = self._get_pk_val(meta)
         if not self._is_pk_set(meta):
             pk_val = meta.pk.get_pk_value_on_save(self)
             setattr(self, meta.pk.attname, pk_val)
@@ -1129,6 +1128,7 @@ class Model(AltersData, metaclass=ModelBase):
                 for f in non_pks_non_generated
             ]
             forced_update = update_fields or force_update
+            pk_val = self._get_pk_val(meta)
             updated = self._do_update(
                 base_qs, using, pk_val, values, update_fields, forced_update
             )
```

File: django/db/models/fields/related.py
Comment (nessita) [inline django/db/models/fields/related.py:1970]: I imagined you wanted to minimize the diff but I think the following may be preferable:
```suggestion
        return list(getattr(obj, self.attname).all()) if obj._is_pk_set() else []
```

File: docs/ref/models/instances.txt
Comment (nessita) [inline docs/ref/models/instances.txt:982]:
```suggestion
The ``_is_pk_set()`` method returns whether the ``pk`` property is defined or not.
```

File: docs/ref/models/instances.txt
Comment (nessita) [inline docs/ref/models/instances.txt:984]: Do we need to clarify this? (honest question)
In a way, we are adding this helper to abstract the users from knowing/understanding when a pk is set, so maybe we should remove it (perhaps this is a case where "less is more").
What do you think?

File: docs/releases/5.2.txt
Comment (nessita) [inline docs/releases/5.2.txt:286]: This needs to be removed.

File: docs/releases/5.2.txt
Comment (nessita) [inline docs/releases/5.2.txt:292]: Given that this is still an "experimental" method, I would make this note more succinct:
```suggestion
* The new :meth:`~django.db.models.Model._is_pk_set()` method allows to
  check whether a Model instance has its primary key set or not.
```

**Review by nessita — COMMENTED** [review]
Thank you @csirmazbendeguz! This is almost there! :trophy: 

I added a few comments and I have this "philosophical" concern after close code inspection:

There are so many occurrences of `if NOT obj._is_pk_set()` (in fact the majority of this diff is about checking "is not set" vs "is set) which makes me wonder if we should also provide:

```python
def _is_pk_unset(self):
    return self.pk is None  # pseudocode
```

thinking also in the future implementation for composite PKs being (note usage of `all` vs `any`):

```python
def _is_pk_unset(self):
    return any(i is None for i in self.pk)  # pseudocode
```

which would not require to check every element in the whole tuple, just to negate the result in the call site. Does this make sense? What do you think?

**Comment by collinanderson** [comment]
I'm also realizing it could be more clear to call this method `all_pk_set()` to be clear it's not `any_pk_set()` but kinda bikeshedding at some point.

**Comment by csirmazbendeguz** [comment]
Thanks @nessita !

This is just my opinion but `_is_pk_set` reads better.

> which would not require to check every element in the whole tuple

`all(i is not None for i in self.pk)` wouldn't require checking every element either.

**Review by nessita — APPROVED** [review]
Thank you @csirmazbendeguz for this work and your patience bearing with us in the developing of the final solution. :star2: 

I have pushed minor edits and I will merge once CI is green. Onward to the composite PK PR!

**Comment by csirmazbendeguz** [comment]
Thanks @nessita ! This is great news. 🎉 I rebased the [CompositePrimaryKey patch](https://github.com/django/django/pull/18056), it's ready for review. This is the PR that's going to lay the foundation for composite primary keys in Django. I'm not sure if you're gonna have the time to review it though, it's quite a large patch. But any feedback is appreciated if someone has the time here.

**Pull request merged**
